File: src/main/java/core/db/TournamentDetailsTable.java
package core.db;

import core.model.Tournament.TournamentDetails;
import core.util.HODateTime;
import core.util.HOLogger;

import java.sql.ResultSet;
import java.sql.Types;

import static core.util.DateTimeUtils.DateToSQLtimeStamp;

public final class TournamentDetailsTable extends AbstractTable {

	public final static String TABLENAME = "TOURNAMENTDETAILS";

	protected TournamentDetailsTable(JDBCAdapter  adapter){
		super(TABLENAME,adapter);
	}

	@Override
	protected void initColumns() {
		columns = new ColumnDescriptor[14];
		columns[0]= new ColumnDescriptor("TournamentId",Types.INTEGER,false,true); //The globally unique identifier of the match
		columns[1]= new ColumnDescriptor("Name",Types.VARCHAR,false,256);
		columns[2]= new ColumnDescriptor("TournamentType",Types.INTEGER,false);
		columns[3]= new ColumnDescriptor("Season",Types.SMALLINT,false);  //Current season
		columns[4]= new ColumnDescriptor("LogoUrl",Types.VARCHAR,true,256); //The url to the logo
		columns[5]= new ColumnDescriptor("TrophyType",Types.INTEGER,false);
		columns[6]= new ColumnDescriptor("NumberOfTeams",Types.INTEGER,false);
		columns[7]= new ColumnDescriptor("NumberOfGroups",Types.INTEGER,false);
		columns[8]= new ColumnDescriptor("LastMatchRound",Types.TINYINT,false);
		columns[9]= new ColumnDescriptor("FirstMatchRoundDate",Types.TIMESTAMP,false);
		columns[10]= new ColumnDescriptor("NextMatchRoundDate",Types.TIMESTAMP,true); //The date of the next match round. During matches this will be the Next match round.
		columns[11]= new ColumnDescriptor("IsMatchesOngoing",Types.BOOLEAN,false);
		columns[12]= new ColumnDescriptor("Creator_UserID",Types.INTEGER,false);
		columns[13]= new ColumnDescriptor("Creator_Loginname",Types.VARCHAR,true, 256);
	}

	@Override
	protected PreparedSelectStatementBuilder createPreparedSelectStatementBuilder(){
		return new PreparedSelectStatementBuilder(this," WHERE TOURNAMENTID = ?" );
	}

	public TournamentDetails getTournamentDetails(int tournamentId)
	{
		TournamentDetails oTournamentDetails = null;
		try {
			var rs = executePreparedSelect(tournamentId);
			assert rs != null;
			if (rs.next()) {
				oTournamentDetails = new TournamentDetails();
				oTournamentDetails.setTournamentId(rs.getInt("TournamentId"));
				oTournamentDetails.setName(rs.getString("Name"));
				oTournamentDetails.setTournamentType(rs.getInt("TournamentType"));
				oTournamentDetails.setSeason((short)rs.getInt("Season"));
				oTournamentDetails.setLogoUrl(rs.getString("LogoUrl"));
				oTournamentDetails.setTrophyType(rs.getInt("TrophyType"));
				oTournamentDetails.setNumberOfTeams(rs.getInt("NumberOfTeams"));
				oTournamentDetails.setNumberOfGroups(rs.getInt("NumberOfGroups"));
				oTournamentDetails.setLastMatchRound((short)rs.getInt("LastMatchRound"));
				oTournamentDetails.setFirstMatchRoundDate(HODateTime.fromDbTimestamp(rs.getTimestamp("FirstMatchRoundDate")));
				oTournamentDetails.setFirstMatchRoundDate(HODateTime.fromDbTimestamp(rs.getTimestamp("NextMatchRoundDate")));
				oTournamentDetails.setMatchesOngoing(rs.getBoolean("IsMatchesOngoing"));
				oTournamentDetails.setCreator_UserId(rs.getInt("Creator_UserID"));
				oTournamentDetails.setCreator_Loginname(rs.getString("Creator_Loginname"));
			}
		}
		catch (Exception e) {
			HOLogger.instance().log(getClass(),
					"DB.getTournamentDetails Error" + e);
		}
		return oTournamentDetails;
	}


	/**
	 * Store Tournament Details into DB
	 */
	void storeTournamentDetails(TournamentDetails details) {
		StringBuilder sql = new StringBuilder(200);

		try {
			executePreparedInsert(
					details.getTournamentId(),
					details.getName(),
					details.getTournamentType(),
					details.getSeason(),
					details.getLogoUrl(),
					details.getTrophyType(),
					details.getNumberOfTeams(),
					details.getNumberOfGroups(),
					details.getLastMatchRound(),
					HODateTime.toDbTimestamp(details.getFirstMatchRoundDate()),
					HODateTime.toDbTimestamp(details.getNextMatchRoundDate()),
					details.getMatchesOngoing(),
					details.getCreator_UserId(),
					details.getCreator_Loginname()
			);
		} catch (Exception e) {
			HOLogger.instance().log(getClass(),
					"DB.storeTournamentDetails Error" + e);
			HOLogger.instance().log(getClass(), e);
		}

	}

}

File: src/main/java/core/db/TournamentDetailsTable.kt
package core.db

import core.model.Tournament.TournamentDetails
import core.util.HODateTime
import java.sql.*
import java.util.function.BiConsumer
import java.util.function.Function

class TournamentDetailsTable internal constructor(adapter: JDBCAdapter) : AbstractTable(TABLENAME, adapter) {
    override fun initColumns() {
        columns = arrayOf<ColumnDescriptor>(
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("TournamentId")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.tournamentId }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.tournamentId = v as Int })
                .setType(
                    Types.INTEGER
                ).isNullable(false).isPrimaryKey(true).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("Name")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.name }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any? -> (p as TournamentDetails?)!!.name = v as String? })
                .setType(Types.VARCHAR).setLength(256).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("TournamentType")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.tournamentType }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.tournamentType = v as Int })
                .setType(
                    Types.INTEGER
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("Season")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.season }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.season = (v as Int).toShort() })
                .setType(
                    Types.INTEGER
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("LogoUrl")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.logoUrl }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any? -> (p as TournamentDetails?)!!.logoUrl = v as String? })
                .setType(
                    Types.VARCHAR
                ).setLength(256).isNullable(true).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("TrophyType")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.trophyType }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.trophyType = v as Int }).setType(
                Types.INTEGER
            ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("NumberOfTeams")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.numberOfTeams }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.numberOfTeams = v as Int })
                .setType(
                    Types.INTEGER
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("NumberOfGroups")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.numberOfGroups }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.numberOfGroups = v as Int })
                .setType(
                    Types.INTEGER
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("LastMatchRound")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.lastMatchRound }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any ->
                    (p as TournamentDetails?)!!.lastMatchRound = (v as Int).toShort()
                }).setType(
                Types.INTEGER
            ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("FirstMatchRoundDate")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.firstMatchRoundDate.toDbTimestamp() })
                .setSetter(
                    BiConsumer<Any?, Any> { p: Any?, v: Any? ->
                        (p as TournamentDetails?)!!.firstMatchRoundDate = v as HODateTime?
                    }).setType(
                Types.TIMESTAMP
            ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("NextMatchRoundDate")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.nextMatchRoundDate.toDbTimestamp() })
                .setSetter(
                    BiConsumer<Any?, Any> { p: Any?, v: Any? ->
                        (p as TournamentDetails?)!!.nextMatchRoundDate = v as HODateTime?
                    }).setType(
                Types.TIMESTAMP
            ).isNullable(true).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("IsMatchesOngoing")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.matchesOngoing }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any -> (p as TournamentDetails?)!!.matchesOngoing = v as Boolean })
                .setType(
                    Types.BOOLEAN
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("Creator_UserID")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.creator_UserId }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any? -> (p as TournamentDetails?)!!.creator_UserId = v as Int? })
                .setType(
                    Types.INTEGER
                ).isNullable(true).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("Creator_Loginname")
                .setGetter(Function<Any?, Any?> { p: Any? -> (p as TournamentDetails?)!!.creator_Loginname }).setSetter(
                BiConsumer<Any?, Any> { p: Any?, v: Any? ->
                    (p as TournamentDetails?)!!.creator_Loginname = v as String?
                }).setType(
                Types.VARCHAR
            ).setLength(256).isNullable(true).build()
        )
    }

    fun getTournamentDetails(tournamentId: Int): TournamentDetails? {
        return loadOne(TournamentDetails::class.java, tournamentId)
    }

    /**
     * Store Tournament Details into DB
     */
    fun storeTournamentDetails(details: TournamentDetails?) {
        store(details)
    }

    companion object {
        const val TABLENAME = "TOURNAMENTDETAILS"
    }
}
